### ISSUE TITLE
OpenSSLCertificateOptions contains asserts

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6288]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)| @hynek reported|
|-|-|
|Trac ID|trac#6288|
|Type|defect|
|Created|2013-01-31 09:29:56Z|
|Branch|https://github.com/twisted/twisted/tree/remove-asserts-6288|

There are two `assert`s in the constructor of `twisted.internet._sslverify.OpenSSLCertificateOptions`.

Additionally the second `assert` is not in sync with the documentation which says for `caCerts`:

```
Only used if verify is True, and if verify is True, this must be specified.
```

It checks only for `caCerts` being set if `verify` is `True` though.

This also begs the question what the `verify` option is good for anyway. Interestingly, the only place where `self.verify` is used is `_makeContext` and `caCert` is checked before usage:

```python
        if self.verify:
            verifyFlags = SSL.VERIFY_PEER
            if self.requireCertificate:
                verifyFlags |= SSL.VERIFY_FAIL_IF_NO_PEER_CERT
            if self.verifyOnce:
                verifyFlags |= SSL.VERIFY_CLIENT_ONCE
            if self.caCerts:
                store = ctx.get_cert_store()
                for cert in self.caCerts:
                    store.add_cert(cert)

```

So either the docs are wrong or this code can be simplified.

I’ll attach a proposed patch which assumes the docs to be correct thus making it possible to optimize that `if` away and tighten the parameter check. The tests don't mind.

Attachments:

* [remove-asserts-6288.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/929/92952761ba915e715faf4cdec4c61df4fadfddd1/3970d5cc2516b8f040170ec8fbb574accd7989a5.patch) (7093 bytes) - added by hynek on 2013-01-31 09:33:18Z - 
* [remove-asserts-6288-v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/929/92952761ba915e715faf4cdec4c61df4fadfddd1/af35c9c5deaa8e040a026a22d8af472109eeda5c.patch) (11452 bytes) - added by hynek on 2013-02-02 16:21:48Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6288 6288
type__defect defect
reporter__hynek hynek
priority__normal normal
milestone__ 
branch__branches_remove_asserts_6288 branches/remove-asserts-6288
branch_author__hynek hynek
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1359624596000000 1359624596000000
changetime__1360880498000000 1360880498000000
version__None None
owner__tom_prince tom.prince
cc__hynek
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @hynek|
|-|-|

Thanks.

From the description: 

> So either the docs are wrong or this code can be simplified.

This doesn't seem right to me.  The logic seemed to make sense before, and the documentation agreed with the implementation (at least the non-assert part of the implementation).

If it helps, here's a re-phrasing.  If verify is False, caCerts will be ignored.  If verify is True, if a caCerts value is given, it will be used.

Apart from that, it's better if test methods only make one assertion (so you can see all the problems after a single test run), and the last statement of `test_constructorEnforcesNeitherOrBothCaCertsAndVerify` should be an assertion, not creation of an object that is then thrown away without being inspected at all.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_3"></a>|@hynek removed owner|
|-|-|

Thanks for the feedback!

I've attached another attempt which also clarifies the doc string a bit – and matches the style of verify.  It’s purely syntax in the latter case though (and the reflow is has been caused by adding proper epydoc markup, not the indent).

Let me know what you think.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_4"></a>|@tomprince commented|
|-|-|

(In [37135]) Branching to remove-asserts-6288.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince set owner to @tomprince|
|-|-|

This looks good. I'll merge.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince set status to `closed`|
|-|-|

(In [37139]) Merge remove-asserts-6288: Remove asserts from OpenSSLCertificateOptions.

Author: hynek
Reviewers: exarkun, tom.prince
Fixes: [#6288](https://github.com/twisted/twisted/issues/6288)

This replaces them with raising ValueError in the appropriate cases. This
avoids the checks being removed when running with `-O`.
